src/Nethermind/Nethermind.Mev/Source/BundlePool.cs

```diff
@@ -61,60 +88,74 @@ public BundlePool(IBlockTree blockTree, IBundleSimulator simulator, IBlockFinali
 
         private IEnumerable<MevBundle> GetBundles(long blockNumber, UInt256 minTimestamp, UInt256 maxTimestamp, CancellationToken token = default)
         {
-            int CompareBundles(MevBundle searchedBundle, KeyValuePair<MevBundle, ConcurrentBag<Keccak>> potentialBundle)
+            if (_bundles.TryGetBucket(blockNumber, out MevBundle[] bundles))
             {
-                return searchedBundle.BlockNumber <= potentialBundle.Key.BlockNumber ? -1 : 1;
-            }
-
-            lock (_bundles)
-            {
-                MevBundle searchedBundle = MevBundle.Empty(blockNumber, minTimestamp, maxTimestamp);
-                int i = _bundles.BinarySearch(searchedBundle, CompareBundles);
-                for (int j = (i >= 0 ? i : ~i); j < _bundles.Count; j++)
+                foreach (MevBundle mevBundle in bundles)
                 {
                     if (token.IsCancellationRequested)
                     {
                         break;
                     }
 
-                    MevBundle mevBundle = _bundles[j].Key;
-                    if (mevBundle.BlockNumber == searchedBundle.BlockNumber)
-                    {
-                        bool bundleIsInFuture = mevBundle.MinTimestamp != UInt256.Zero && searchedBundle.MinTimestamp < mevBundle.MinTimestamp;
-                        bool bundleIsTooOld = mevBundle.MaxTimestamp != UInt256.Zero && searchedBundle.MaxTimestamp > mevBundle.MaxTimestamp;
-                        if (!bundleIsInFuture && !bundleIsTooOld)
-                        {
-                            yield return mevBundle;
-                        }
-                    }
-                    else
+                    bool bundleIsInFuture = mevBundle.MinTimestamp != UInt256.Zero && minTimestamp < mevBundle.MinTimestamp;
+                    bool bundleIsTooOld = mevBundle.MaxTimestamp != UInt256.Zero && maxTimestamp > mevBundle.MaxTimestamp;
+                    if (!bundleIsInFuture && !bundleIsTooOld) 
                     {
-                        break;
+                        yield return mevBundle;
                     }
                 }
             }
+                    
         }
 
         public bool AddBundle(MevBundle bundle)
+        {
+            if (ValidateBundle(bundle))
+            {
+                bool result;
+
+                lock (_bundles)
+                {
+                    result = _bundles.TryInsert(bundle, bundle);
+                }
+
+                if (result)
+                {
+                    SimulateBundle(bundle);
```

Comment:
do you simulate bundles on adding? what if the block is not yet known?

Reply:
fixed